casper/src/main/resources/RevVault.rho

```diff
@@ -27,255 +28,242 @@ in {
-                @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
-                @Either!("flatMap <-", *revAddressValidEitherCh, *getOrCreate, *retCh) |
-                for (_, ret <- getOrCreate) {
-                  // At first peek for `vaultMap` to check if vault (address) exists
-                  // - peek ensures that for existing vaults we don't have a block conflict
-                  for (@vaultMap <<- vaultMapStore) {
-                    match vaultMap.get(revAddress) {
-                      Nil => {
-                        // Vault is not found, create new and save it in the `vaultMap`
-                        // - this means two blocks with new vaults cannot be merged
-                        for (@vaultMap <- vaultMapStore) {
-                          // We need to check again because `vaultMap` is not locked with peek
-                          match vaultMap.get(revAddress) {
-                            Nil => {
-                              new resCh in {
-                                constructor!(*resCh) |
-                                for (@eitherVault <- resCh) {
-                                  ret!(eitherVault) |
-                                  match eitherVault {
-                                    (true, vault) => vaultMapStore!(vaultMap.set(revAddress, vault))
-                                    (false, _)    => vaultMapStore!(vaultMap)
-                                  }
+            }
+          } |
+
+          contract _getOrCreate(@revAddress, constructor, retCh, ackCh) = {
+            new revAddressValidCh, revAddressValidEitherCh in {
+              RevAddress!("validate", revAddress, *revAddressValidCh) |
+              @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
+              for (@revAddressEither <- revAddressValidEitherCh) {
+                match revAddressEither {
+                  // Invalid Rev address
+                  (false, _) => {
+                    retCh!(revAddressEither) |
+                    ackCh!(Nil)
+                  }
+                  _ => {
+                    // Valid Rev address
+                    // "get" returns Nil if vault is not in the store, otherwise it returns the vault.
+                    new vaultCh in {
+                      TreeHashMap!("get", vaultMap, revAddress, *vaultCh) |
+                      for (@vault <- vaultCh) {
+                        match vault {
+                          Nil => {
+                            // The vault was not found, create a new one and set it in the vaultMap.
+                            new resCh in {
+                              constructor!(*resCh) |
+                              for (@eitherVault <- resCh) {
+                                retCh!(eitherVault) |
+                                match eitherVault {
+                                  (true, vault) => { TreeHashMap!("set", vaultMap, revAddress, vault, *ackCh) }
```

Comment:
Vault is returned before it's `set` to vaults map. Can we just wait on TreeMap result and then return vault on `retCh`?

Also `get` on TreeMap is not blocking. I'm not sure if we need more synchronization. Maybe `update` would be better here? Between get and set other thread can create new vault for the same address.

Reply:
I changed the the vault return to occur after the acknowledgement from `set`.